docs/2.10.0/docs/concepts/ledger-model/upgrade-model-in-depth.rst

```diff
@@ -0,0 +1,1680 @@
+
+     - .. code-block:: haskell
+
+            data T = 
+              A Int | B Text | C Bool
+
+Below, the variant on the right is **not** a valid upgrade of the
+variant on the left because it adds a new constructor ``C`` before ``B`` instead
+of adding it at the end of the constructor sequence.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: haskell
+
+            data T =
+              A Int | B Text
+
+     - .. code-block:: haskell
+
+            data T = 
+              A Int | C Bool | B Text
+
+Below, the variant on the right is **not** a valid upgrade of the
+variant on the left because it changes the order of its constructors.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: haskell
+
+            data T =
+              A Int | B Text
+
+     - .. code-block:: haskell
+
+            data T = 
+              B Text | A Int
+
+Below, the variant on the right is **not** a valid upgrade of the
+variant on the left because it drops constructor ``B````.``
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: haskell
+
+            data T =
+              A Int | B Text
+
+     - .. code-block:: haskell
+
+            data T = 
+              A Int
+
+Below, the variant on the right is **not** a valid upgrade of the
+variant on the left because it changes the type of ``B``'s argument from
+``Text`` to ``Bool``. ``Bool`` is not a valid upgrade of ``Text``.
+
+.. list-table::
+   :widths: 50 50
+   :width: 100%
+   :class: diff-block
+
+   * - .. code-block:: haskell
+
+            data T =
+              A Int | B Text
+
+     - .. code-block:: haskell
+
+            data T = 
+              A Int | B Bool
+
+Data Types - Enums
+~~~~~~~~~~~~~~~~~~
+
+For the purpose of upgrade validation, enums can be treated as a special
+case of variants. The rules of `the section on
+variants <#data-types---variants>`__ apply, only without constructor
+arguments.
+
+Data Types - Type References
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A type reference is an identifier that resolves to a type. For instance,
+consider the following module definitions, from two different packages:
+
+.. code:: daml
+
+  -- In package q
+  module Dep where
+
+  data U = U with x : Int
+  type A = U
+
+.. code:: daml
+
+  -- In package p
+  module M where
+  import qualified Dep
+
+  data T = T with x : Dep.A
+
+In the definition of ``T``, ``Dep.A`` is a type reference that resolves to the
+type with qualified name ``Dep.U`` in package ``q``.
+
+A reference *r2* to a data type upgrades a reference *r1* to a data type
+if and only if:
+
+-  *r2* resolves to a type *t2* with qualified name *q2* in package *p2;*
+-  *r1* resolves to a type *t1* with qualified name *q1* in package *p1;*
+-  The qualified names *q2* and *q1* are the same;
+-  If the LF versions of both *p1* and *p2* support upgrades then
```

Comment:
```suggestion
-  The LF versions of both *p1* and *p2* support upgrades and
```

Reply:
At least that's how I expect the check to work, as otherwise LF 1.15 types could be upgrades of each other.

Reply:
I meant to say:
```
if (supports_upgrades(p1.version) && supports_upgrades(p2.version))
  then p2 `upgrades` p1
  else p1.pkgid == p2.pkgid
```
The 1.15 case is covered by the else branch.

I feel like this is what the sentence is saying. Maybe the "else" isn't obvious?

Reply:
Changed for an either...or to make it clearer.